### ISSUE TITLE
Plugin Example - Open Feedback

### ISSUE THREAD

**Issue by Codeusa** [AurousGroup/Aurous-SDK#1]
Hi everyone, Andrew from Aurous here. Today I wanted to get your feedback on the basic structure of how writing Aurous plugins will work, if you have a suggestion for change or feedback let us know and we can try and make a plugin system that makes it easy to develop.

A standard plugin looks like this

```javascript
activePlugins.helloWorld= {
    name: "",
    version: 0,
    desc: "",
    author: "",
    twitter: "",
    source: "",
    status: false,
     init: function(pluginName, pluginVersion, pluginDesc, pluginAuthor, pluginTwitter, pluginSource, loadStatus) {
        this.name = pluginName;
        this.version = pluginVersion;
        this.desc = pluginDesc;
        this.author = pluginAuthor;
        this.twitter = pluginTwitter;
        this.source = pluginSource;
        this.status = loadStatus;
        this.customFunction();
    },
    customFunction: function() {
      api.writeFile("C:/", contents);
    },
    getName: function() {
        return this.name;
    },
    getVersion: function() {
        return this.version;
    },
    getDescription: function() {
        return this.desc;
    },
    getAuthor: function() {
        return this.author;
    },
    getTwitter: function() {
        return this.twitter;
    },
    getSource: function() {
        return this.source;
    },
    getStatus: function() {
        return this.status;
    }
};
```
All plugins must contain an init function and are populated from a yourpluginnamespace.config (helloWorld.config)

this is your "main" function and you write code from there. 

Since Aurous uses HTML5 you can manipulate it like any other webpage, however you'll have access to SDK functions for example writing a file to disk.

if you would like to specify how plugins and API data should be written please let us know.

you can also help optimize the plugin loader which is located 

https://github.com/AurousGroup/Aurous-SDK/blob/master/loader/pluginLoader.js

**Comment by alxhotel** [comment]
Can the user edit the plugin's settings ?

**Comment by jorda0mega** [comment]
by this “Since Aurous uses HTML5 you can manipulate it like any other webpage” do you mean using plain javascript or can the developer import external libraries for manipulating the dom?

**Comment by Codeusa** [comment]
@alxhotel you mean code? Theoretically, yes. 

@jorda0mega since it can run any Javascript a library would be possible to include

**Comment by jorda0mega** [comment]
just thinking that it would be sweet to leverage react to write custom components...

**Comment by Codeusa** [comment]
@jorda0mega we will add whatever people come up with for building new components for Aurous, trying to make it as easy as possible to interface with core code

**Comment by fonix232** [comment]
Why not make it a bit more structured, and move all the static information of the plugin (name, version, author, etc.) into a separate json file? That way, plugins can be loaded a bit faster, plus the information is in a format that was meant to hold static info (it would also help defining permissions, code entry points, menu entries, etc., basically things that are just in the way if it's placed next to the functionality).

**Comment by Codeusa** [comment]
Hi @fonix232 

All the information is populated from a json file, there are only predefined functions so they are viewable in the namespace without constantly reloading a file.

If you want to provide an example of your suggestion we can take a look at implementing it.

**Comment by fonix232** [comment]
As you requested, an example:

https://gist.github.com/fonix232/5ea6a7f2f5414413016b

**Comment by Pappa** [comment]
The getters aren't serving any practical purpose because the properties they give access to aren't private.

**Comment by Promises** [comment]
This will be great for conky :)

**Comment by lricher** [comment]
Will there be a platform to share/sell plugins we write for Aurous?

**Comment by Codeusa** [comment]
@lricher yes we plan to allow people to sell their plugins and accept donations.

**Comment by fonix232** [comment]
@Codeusa I'm not sure a payware system for plugins for an app that pretty much allows downloading free music will be very welcome :stuck_out_tongue:

**Comment by Draxler** [comment]
@Codeusa 
I agree with @fonix232. Introducing payware system for plugins have a litlle sence (to me anyway). Imo, in this case Aurous devs could be in a lot of trouble, the moment people (without copyrights) start asking others for money to exces music. And this time it wouldn't mether if it's third party plugin to copyrights owners. Copyright holders will saw Aurous devs acountable once payment starts. Aurous dev(s) should think hard about payware system for plugins.

In case Aurous go ahead with payware system for plugins anyway: 
Would Aurous welcomed free plugins also into Aurous app or is there a plan to only introduce payware system for plugins?

**Comment by Codeusa** [comment]
@fonix232  No one is entitled to a plugin another developer creates, if a developer uses our SDK to create a plugin which they won't to sell on our platform, they're free to. Aurous is a music player. Our search functionality has nothing to do with what is "welcome"

**Comment by fonix232** [comment]
@Codeusa Let's see this from another angle. You're downloading an app to
access music for free (a la Popcorn Time). Now imagine that there are
plugins, but to download them, you have to pay. Would you? You're
downloading this app to basically avoid paying for something (which is
understandable in certain situations, e.g. when you're not harming anyone
with it). So when a way to avoid paying requires you to pay for a part of
it, you'll discard it.

I'm not saying it does not make sense to make paid addons, but it will
drive possible users away. So my suggestion is  that you leave the
licensing of the plugins to the creators, but you won't operate a store of
any kind for paid plugins - that way you avoid any possible legal
consequences, and also avoid losing users. So make the official store of
plugins FOSS license compliant: any plugin that is free and open source,
will be accepted and directly downloaded to the user from Git.

This also brings another question - how will be paid plugins distributed?
As they are simple JS files, how do you make sure they are not easily
redistributed, for free? How will you protect your paid content from the
exact kind of actions you're enabling your users to do with music? Take
uTorrent for example (and they're a native app, so easier to encrypt and
secure): all their premium releases end up on torrent trackers, because
people want the extra features for free - as the app enables them to get
everything for pretty much free, they expect the app itself to be free with
all the features. I doubt uTorrent as a company even makes much money on
pure sales, and rather monetizes on the free versions ads.

Now take Aurous -  it's a wholly open source project written in a language
where you can't really encrypt code in a secure way, and all the "paid
features" are actually extensions written by a third party. Do you  REALLY
think this will be a sustainable model?

**Comment by Codeusa** [comment]
@fonix232 you're arguing this as if we plan to take a share, we don't.

**Comment by fonix232** [comment]
I'm not talking about shares or percentages. I'm talking about prestige and
user experience. I didn't even consider your financial partaking in the
whole thing.

Of course, let the developers decide if they want to make their plugins
paid. But I think it is a bad idea in general to promote paid extensions of
an app that itself makes it possible to avoid paying for content. It's a
bit controversial, and rises questions about intentions.

I do not want to offend you in any way, but in the future, the Aurous team
should think before making any kind of statement - as the wording, or
implied meaning of these can mean a lot to the users. Hell, I'm all in for
you guys putting ads into the app if that helps with the development and is
non-intrusive (no audio ads, no popups, no "listen to this ad before next
track" kind of bullshit, so pretty much unlike Spotify). You making a few
bucks off of happy users indirectly won't make me upset, you're spending
time on it. But people SELLING plugins that will be most likely used to
access certain contents possibly in a license-breaking way is not ethical
in my opinion.

But let's just agree to disagree. I for one do not find endorsing paid
plugins for this kind of application either ethical nor a viable business
plan. I voiced my concern, and you've taken it as me attacking your
business plan - I did not, I just tried to point out that in my opinion,
it's flawed. However I do hope that you succeed and that Aurous will be a
success story, and not just app that is tagged as a "short lived fad".

**Comment by Codeusa** [comment]
@fonix232 I should make it clear, we will not allow infringing plugins on our platform, we will adhere to DMCA when it comes to plugins that allow for activity we see unfit

**Comment by lricher** [comment]
@Codeusa Thanks for the info!

I think its a great idea to leverage your plugin system in this way. Free or paid, it doesn't matter to me. I'm just happy to see that there will be a platform for sharing plugins!

**Comment by gubatron** [comment]
regarding the structure of plugins is there a list of "hooks" upon which all plugins are called?
say music playback related hooks: `onAudioPlay(...)`, `onAudioPause()`, `onStop()`
search related hooks (in case you wanted to extend the search functionality to connect to other networks), app life cycle hooks, etc. ?
